File: aoc_2022_javascript/day03.js
main();

function main() {
  const fs = require("fs");
  const lines = fs
    .readFileSync("input/day03.txt", { encoding: "utf-8" })
    .split("\n")
    .map(line => line.trim());
  let start = Date.now();
  solution = solve(lines);
  let end = Date.now();
  console.log(`Execution time: ${end - start} ms`);
  console.log(`Part 1: ${solution[0]}`);
  console.log(`Part 2: ${solution[1]}`);
}

function solve(lines) {
  let part1 = 0;
  for (let i = 0; i < lines.length; i++) {
    let middle = lines[i].length / 2;
    let items = find_common([lines[i].slice(0, middle), lines[i].slice(middle)]);
    part1 += get_score(items);
  }

  let part2 = 0;
  for (let i = 0; i < lines.length - 2; i+=3) {
    let items = find_common([lines[i], lines[i + 1], lines[i + 2]])
    part2 += get_score(items);
  }
  
  return [part1, part2];
}

function find_common(buckets) {
  let items = new Set();
  for (let i = 0; i < buckets[0].length; i++) {
    let item = buckets[0][i];
    for (let a = 1; a < buckets.length; a++) {
      if (buckets[a].includes(item)) {
        items.add(item);
      }
    }
  }
  return items;
}

function get_score(items) {
  // Lowercase item types a through z have priorities 1 through 26.
  // Uppercase item types A through Z have priorities 27 through 52.
  let ret = 0;
  for (let value of items) {
    let ascii = value.charCodeAt(0);
    if (ascii >= 97) { // lower case
      ret += ascii - 97 + 1;
    } else { // upper case
      ret += ascii - 65 + 27;
    }
  }
  return ret;
}

File: aoc_2024_python/day03.py
import aoc
import re

def main(input: str):
    part1 = sum([int(m["a"]) * int(m["b"]) for m in re.finditer(r"mul\((?P<a>\d+),(?P<b>\d+)\)", input)])

    part2 = 0
    do = True
    for m in re.finditer(r"(?P<do>do\(\))|(?P<dont>don't\(\))|(mul\((?P<a>\d+),(?P<b>\d+)\))", input):
        if m["do"]:
            do = True
        elif m["dont"]:
            do = False
        elif do:
            part2 += int(m["a"]) * int(m["b"])

    return part1, part2

aoc.run_raw(main, "day03.txt")
